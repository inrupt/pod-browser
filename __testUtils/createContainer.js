/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  createSolidDataset,
  createThing,
  addDatetime,
  addDecimal,
  addInteger,
  setThing,
} from "@inrupt/solid-client";
import { rdf, ldp } from "rdf-namespaces";

export const TIMESTAMP = new Date(Date.UTC(2020, 5, 2, 15, 59, 21));

export default function createContainer(
  url = "https://user.dev.inrupt.net/",
  type = ldp.BasicContainer
) {
  const thing = createThing({ url });
  const publicContainer = [
    (t) => addUrl(t, rdf.type, type),
    (t) => addUrl(t, rdf.type, ldp.Container),
    (t) => addDatetime(t, "http://purl.org/dc/terms/modified", TIMESTAMP),
    (t) =>
      addDecimal(t, "http://www.w3.org/ns/posix/stat#mtime", 1591131561.195),
    (t) => addInteger(t, "http://www.w3.org/ns/posix/stat#size", 4096),
    (t) => addUrl(t, ldp.contains, "https://user.dev.inrupt.net/public/games/"),
  ].reduce((acc, fn) => fn(acc), thing);

  return setThing(createSolidDataset(), publicContainer);
}
