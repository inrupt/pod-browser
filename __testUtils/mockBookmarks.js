/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addDatetime,
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { dct, rdf } from "rdf-namespaces";
import {
  RECALLS_PROPERTY_IRI,
  BOOKMARK_TYPE_IRI,
} from "../src/solidClientHelpers/bookmarks";

const bookmarksUrl = "http://example.com/bookmarks/index.ttl";
const bookmark1Title = "Cat gifs";
const bookmark1Recalls = "https://example.com/catgifs";
const bookmark2Title = "Dog videos";
const bookmark2Recalls = "https://example.com/dogvideos";

export default function mockBookmarks() {
  let bookmarksDataset = mockSolidDatasetFrom(bookmarksUrl);
  const bookmark1 = mockThingFrom(`${bookmarksUrl}#1234`);
  const bookmark1WithTitle = addStringNoLocale(
    bookmark1,
    dct.title,
    bookmark1Title
  );
  const bookmark1WithRecalls = addUrl(
    bookmark1WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark1Recalls
  );
  const bookmark1WithType = addUrl(
    bookmark1WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark1WithCreated = addDatetime(
    bookmark1WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 28, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark1WithCreated);
  const bookmark2 = mockThingFrom(`${bookmarksUrl}#4567`);
  const bookmark2WithTitle = addStringNoLocale(
    bookmark2,
    dct.title,
    bookmark2Title
  );
  const bookmark2WithRecalls = addUrl(
    bookmark2WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark2Recalls
  );
  const bookmark2WithType = addUrl(
    bookmark2WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark2WithCreated = addDatetime(
    bookmark2WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 29, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark2WithCreated);
  return {
    dataset: bookmarksDataset,
    iri: bookmarksUrl,
  };
}
