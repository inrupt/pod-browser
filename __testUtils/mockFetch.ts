/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import "whatwg-fetch";
import Mock = jest.Mock;

type fetchType = typeof window.fetch;

export default function mockFetch(
  responses: Record<string, Response | Record<string, Response>>
): Mock<fetchType> {
  return jest
    .fn<any, Parameters<fetchType>>()
    .mockImplementation((url, init) => {
      const response = responses[url.toString()];
      const responseObject = response as Record<string, Response>;
      const method = init?.method || "GET";
      if (typeof responseObject[method] !== "undefined") {
        return Promise.resolve(responseObject[method]);
      }
      if (response) {
        return Promise.resolve(response);
      }
      throw new Error(`URL (${url}) not mocked properly`);
    });
}

export function mockSolidClientRequest(
  body: string,
  method = "GET",
  headers: Record<string, string> = {}
): Request {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status: number, body = ""): Response {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
