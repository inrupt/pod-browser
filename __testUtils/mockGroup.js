/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  getStringNoLocale,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/group";
import { vcardExtras } from "../src/models/addressBook";

function mockGroupThing(name, groupThingUrl) {
  return chain(
    mockThingFrom(groupThingUrl),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, vcard.fn, name)
  );
}

function mockIndexThing(addressBook, groupThingUrl) {
  return setUrl(addressBook.thing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
  );
}

// export function addGroupsToAddressBook(addressBook, groupsToAdd) {
//   const { groups, ...rest } = addressBook;
//   return {
//     ...rest,
//     groups: {
//       iri: addressBook.groups.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...groupsToAdd.map(({ dataset, iri }) => (d) => {
//           const group = getThing(dataset, iri);
//           const name = getStringNoLocale(group, vcard.fn);
//           return setThing(d, mockGroupThing(name, iri));
//         }),
//         ...groupsToAdd.map((group) => (d) =>
//           setThing(d, mockIndexThing(addressBook, group.iri))
//         )
//       ),
//     },
//   };
// }

export default function mockGroup(
  addressBook,
  name,
  { url, id, members } = {}
) {
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = chain(
    mockGroupThing(name, groupThingUrl),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    )
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
