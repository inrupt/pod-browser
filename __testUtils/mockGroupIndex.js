/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { mockSolidDatasetFrom } from "@inrupt/solid-client";
import { getContactIndexDefaultUrl } from "../src/models/contact/collection";
import { addGroupToMockedIndexDataset } from "./mockGroupContact";
import { getGroupName, getGroupUrl } from "../src/models/group";
import { chain } from "../src/solidClientHelpers/utils";
import { GROUP_CONTACT } from "../src/models/contact/group";

export default function mockGroupIndex(addressBook, groups) {
  const contactIndexUrl = getContactIndexDefaultUrl(
    addressBook.containerUrl,
    GROUP_CONTACT
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(contactIndexUrl),
      ...groups.map((g) => (d) =>
        addGroupToMockedIndexDataset(
          d,
          addressBook,
          getGroupName(g),
          getGroupUrl(g)
        )
      )
    ),
    type: GROUP_CONTACT,
  };
}
