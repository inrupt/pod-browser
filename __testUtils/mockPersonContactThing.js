/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, foaf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/models/addressBook";

export const webIdUrl = "http://example.com/alice#me";

export function addMockedPersonThingsToIndexDataset(
  dataset,
  addressBook,
  name,
  personThingUrl
) {
  return chain(dataset, (d) =>
    setThing(
      d,
      chain(
        mockThingFrom(personThingUrl),
        (t) => setUrl(t, rdf.type, vcard.Individual),
        (t) => setStringNoLocale(t, vcard.fn, name),
        (t) => setUrl(t, vcardExtras("inAddressBook"), asUrl(addressBook.thing))
      )
    )
  );
}

// export function addContactsToAddressBook(addressBook, contactsToAdd) {
//   // TODO: Should probably be moved to a separate mockContact.js
//   const { people, ...rest } = addressBook;
//   return {
//     ...rest,
//     people: {
//       iri: addressBook.people.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...contactsToAdd.map((contactThing) => (d) =>
//           setThing(
//             d,
//             chain(contactThing, (t) =>
//               setUrl(t, vcardExtras("inAddressBook"), addressBook.index.iri)
//             )
//           )
//         )
//       ),
//     },
//   };
// }

export default function mockPersonContactThing(url = webIdUrl) {
  return chain(
    mockThingFrom(url),
    (t) => addUrl(t, rdf.type, vcard.Individual),
    (t) => addUrl(t, foaf.openid, webIdUrl)
  );
}
