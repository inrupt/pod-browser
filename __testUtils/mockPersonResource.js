/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { addStringNoLocale, addUrl, mockThingFrom } from "@inrupt/solid-client";
import { vcard, foaf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";

export const aliceWebIdUrl = "http://example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockPersonDatasetAlice() {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto)
  );
}

export function mockProfileAlice() {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice());
}

export const bobWebIdUrl = "http://example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";

export function mockPersonDatasetBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick)
  );
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
