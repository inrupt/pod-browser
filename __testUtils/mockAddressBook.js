/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { acl, dc, rdf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { aliceWebIdUrl } from "./mockPersonResource";
import { vcardExtras } from "../src/addressBook";
import { getAddressBookIndexDefaultUrl } from "../src/models/addressBook";
import { joinPath } from "../src/stringHelpers";

export const defaultContainerUrl = "https://user.example.com/contacts";
export const defaultAddressBookDatasetUrl = joinPath(
  defaultContainerUrl,
  "index.ttl"
);
export const defaultAddressBookThingUrl = `${defaultAddressBookDatasetUrl}#this`;

export function mockAddressBookThing(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const owner = options.owner || aliceWebIdUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  const mainIndexUrl = `${indexUrl}#this`;
  return chain(
    mockThingFrom(mainIndexUrl),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, "Contacts")
  );
}

export function mockAddressBookDataset(
  options = {},
  thing = mockAddressBookThing(options)
) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  return chain(mockSolidDatasetFrom(indexUrl), (d) => setThing(d, thing));
}

export default function mockAddressBook(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const thing = mockAddressBookThing(options);
  return {
    containerUrl,
    thing,
    dataset: mockAddressBookDataset(options, thing),
  };
}

export function mockEmptyAddressBook(containerUrl, owner, title = "Contacts") {
  const thing = chain(
    mockAddressBookThing({ containerUrl, owner }),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, title)
  );
  return {
    containerUrl,
    dataset: setThing(createSolidDataset(), thing),
    thing,
  };
}
