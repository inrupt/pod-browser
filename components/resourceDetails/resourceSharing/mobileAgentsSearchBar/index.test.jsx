/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/dom";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import MobileAgentsSearchBar, {
  TESTCAFE_ID_MOBILE_SEARCH_INPUT,
  TESTCAFE_ID_SEARCH_BUTTON,
  TESTCAFE_ID_CLOSE_BUTTON,
} from "./index";

describe("AgentSearchBar", () => {
  const handleFilterChange = jest.fn();
  it("renders a search icon", () => {
    const { asFragment, getByTestId } = renderWithTheme(
      <MobileAgentsSearchBar handleFilterChange={handleFilterChange} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_SEARCH_BUTTON)).toBeDefined();
    expect(getByTestId(TESTCAFE_ID_MOBILE_SEARCH_INPUT)).not.toBeVisible();
  });
  it("clicking the icon expands the search bar", () => {
    const { asFragment, getByTestId } = renderWithTheme(
      <MobileAgentsSearchBar handleFilterChange={handleFilterChange} />
    );
    expect(asFragment()).toMatchSnapshot();
    const button = getByTestId(TESTCAFE_ID_SEARCH_BUTTON);
    userEvent.click(button);
    waitFor(() =>
      expect(getByTestId(TESTCAFE_ID_MOBILE_SEARCH_INPUT)).toBeVisible()
    );
  });
  it("clicking the close icon hides the search bar", () => {
    const { asFragment, getByTestId } = renderWithTheme(
      <MobileAgentsSearchBar handleFilterChange={handleFilterChange} />
    );
    expect(asFragment()).toMatchSnapshot();
    const searchButton = getByTestId(TESTCAFE_ID_SEARCH_BUTTON);
    userEvent.click(searchButton);
    const closeButton = getByTestId(TESTCAFE_ID_CLOSE_BUTTON);
    userEvent.click(closeButton);

    waitFor(() =>
      expect(getByTestId(TESTCAFE_ID_MOBILE_SEARCH_INPUT)).not.toBeVisible()
    );
  });
  it("triggers handleFilterChange when typing into the search input", () => {
    const { getByTestId } = renderWithTheme(
      <MobileAgentsSearchBar handleFilterChange={handleFilterChange} />
    );

    const input = getByTestId(TESTCAFE_ID_MOBILE_SEARCH_INPUT);
    userEvent.type(input, "A");

    expect(handleFilterChange).toHaveBeenCalled();
  });
});
