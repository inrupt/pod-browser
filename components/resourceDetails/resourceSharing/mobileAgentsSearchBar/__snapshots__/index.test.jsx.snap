// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileAgentSearchBar clicking the close icon hides the search bar 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-searchBoxContainer--hidden"
  >
    <button
      aria-label="search"
      class="PodBrowser-root PodBrowser-root"
      data-testid="search-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="search"
          class="PodBrowser-icon-search PodBrowser-iconSearch"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <div
      class="PodBrowser-root PodBrowser-searchInput--hidden PodBrowser-focused"
    >
      <input
        aria-label="Search by name or WebId"
        class="PodBrowser-input PodBrowser-searchInput--hidden"
        data-testid="mobile-search-input"
        placeholder="Search by name or WebId"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="close"
      class="PodBrowser-root PodBrowser-root PodBrowser-iconCloseHidden"
      data-testid="close-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="close"
          class="PodBrowser-icon-cancel hidden"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`MobileAgentSearchBar clicking the icon expands the search bar 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-searchBoxContainer--hidden"
  >
    <button
      aria-label="search"
      class="PodBrowser-root PodBrowser-root"
      data-testid="search-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="search"
          class="PodBrowser-icon-search PodBrowser-iconSearch"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <div
      class="PodBrowser-root PodBrowser-searchInput--hidden PodBrowser-focused"
    >
      <input
        aria-label="Search by name or WebId"
        class="PodBrowser-input PodBrowser-searchInput--hidden"
        data-testid="mobile-search-input"
        placeholder="Search by name or WebId"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="close"
      class="PodBrowser-root PodBrowser-root PodBrowser-iconCloseHidden"
      data-testid="close-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="close"
          class="PodBrowser-icon-cancel hidden"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`MobileAgentSearchBar renders a search icon 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-searchBoxContainer--hidden"
  >
    <button
      aria-label="search"
      class="PodBrowser-root PodBrowser-root"
      data-testid="search-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="search"
          class="PodBrowser-icon-search PodBrowser-iconSearch"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <div
      class="PodBrowser-root PodBrowser-searchInput--hidden PodBrowser-focused"
    >
      <input
        aria-label="Search by name or WebId"
        class="PodBrowser-input PodBrowser-searchInput--hidden"
        data-testid="mobile-search-input"
        placeholder="Search by name or WebId"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="close"
      class="PodBrowser-root PodBrowser-root PodBrowser-iconCloseHidden"
      data-testid="close-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="close"
          class="PodBrowser-icon-cancel hidden"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`MobileAgentSearchBar renders slightly different view with Groups UI enabled 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-searchBoxContainer--hidden PodBrowser-searchBoxContainer--can-be-hidden"
  >
    <button
      aria-label="search"
      class="PodBrowser-root PodBrowser-root"
      data-testid="search-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="search"
          class="PodBrowser-icon-search PodBrowser-iconSearch"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <div
      class="PodBrowser-root PodBrowser-searchInput--hidden PodBrowser-focused"
    >
      <input
        aria-label="Search by name or WebId"
        class="PodBrowser-input PodBrowser-searchInput--hidden"
        data-testid="mobile-search-input"
        placeholder="Search by name or WebId"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="close"
      class="PodBrowser-root PodBrowser-root PodBrowser-iconCloseHidden"
      data-testid="close-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <i
          aria-label="close"
          class="PodBrowser-icon-cancel hidden"
        />
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
